### PR TITLE
[ArgPromotion] Consider InvokeInst in Caller alias analysis

### PR DESCRIPTION
**Pull request by hazzlim**
Check that all users of a Function are CallBase rather than CallInst when performing alias analysis using actual arguments in the calling function, as this check is also valid for Invoke instructions.

This allows replacing the existing check with an assert, as the Function only being used by CallBase derived instructions is a precondition of the transform.

This addresses post-commit review on #106216.

### CODE REVIEW

File: llvm/lib/Transforms/IPO/ArgumentPromotion.cpp [hunk 1]
```diff
@@ -491,10 +491,8 @@ static bool isArgUnmodifiedByAllCalls(Argument *Arg,
                                       FunctionAnalysisManager &FAM) {
   for (User *U : Arg->getParent()->users()) {
 
-    // Bail if we find an unexpected (non CallInst) use of the function.
-    auto *Call = dyn_cast<CallInst>(U);
-    if (!Call)
-      return false;
+    assert(isa<CallBase>(U) && "Expected all users of Function to be CallBase");
+    CallBase *Call = cast<CallBase>(U);
```
Comment (nikic) [inline llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:495]:
```suggestion
    auto *Call = cast<CallBase>(U);
```
`cast` performs an implicit `isa` assert.

**Review by nikic — COMMENTED** [review]
It would be nice to also test that the alias analysis based transform also works with invoked now.

File: llvm/lib/Transforms/IPO/ArgumentPromotion.cpp (hunk 1 above)
Comment (hazzlim) [reply llvm/lib/Transforms/IPO/ArgumentPromotion.cpp]: Nice - updated to remove the superfluous assert(), and added some tests for invoke instructions.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**